node/src/actors/chain_manager/mining.rs

```diff
@@ -313,6 +324,36 @@ impl ChainManager {
                     }
                 })
                 .flatten()
+                // Refrain from trying to resolve any more requests if we have already hit the limit
+                // of retrievals per epoch.
+                .and_then(move |vrf_proof| {
+                    let local_retrieval_count = cloned_retrieval_count.load(Relaxed);
+
+                    if local_retrieval_count + additional_retrieval_count > maximum_retrieval_count {
+                        log::info!("{} Refrained from resolving data request {} for epoch #{} because it contains {} \
+                        sources, which added to the sources that have already been retrieved ({}) would total {} \
+                        retrievals, which exceed current limit per epoch ({}). This limit exists for performance and \
+                        security reasons. You can increase the limit (AT YOUR OWN RISK) by adjusting the \
+                        `data_request_max_retrievals_per_epoch` inside the `[mining]` section in the `witnet.toml` \
+                        configuration file.",
+                            Yellow.bold().paint("[Mining]"),
+                            Yellow.bold().paint(dr_pointer.to_string()),
+                            Yellow.bold().paint(current_epoch.to_string()),
+                            Yellow.bold().paint(additional_retrieval_count.to_string()),
+                            Yellow.bold().paint(local_retrieval_count.to_string()),
+                            Yellow.bold().paint((local_retrieval_count + additional_retrieval_count).to_string()),
+                            Yellow.bold().paint(maximum_retrieval_count.to_string())
+                        );
+
+                        Err(())
+                    } else {
+                        // Update `current_retrieval_count` thanks to interior mutability of the
+                        // `cloned_retrieval_count` reference.
+                        cloned_retrieval_count.fetch_add(additional_retrieval_count, Relaxed);
```

Comment:
There is a race condition here, a different thread could update the value between the load and the store. The simplest way to fix it is to use [compare_and_swap](doc.rust-lang.org/std/sync/atomic/struct.AtomicU16.html#method.compare_and_swap) and retry on error

Reply:
That's an interesting point. I willingly ignored the potential race condition under the assumption that everything works in the same thread, that `current_retrieval_count` is local to the `try_mine_data_request` function and that `try_mine_data_request` is only called once per epoch.

I'll try to incorporate your suggestion just in case my assumptions are proven wrong in the future :+1:

Reply:
Wait a minute, the whole point of using `AtomicU16` is not having to deal with data races:

```rust
pub fn fetch_add(&self, val: $int_type, order: Ordering) -> $int_type {
    // SAFETY: data races are prevented by atomic intrinsics.
    unsafe { atomic_add(self.v.get(), val, order) }
}
```

Reply:
Oh sorry, I was referring to a race condition between this load: `cloned_retrieval_count.load(Relaxed)` and this store: `cloned_retrieval_count.fetch_add`

Reply:
Updated. Feedback please!